### PR TITLE
[Feature] Add UDP TPROXY support for redir on Linux

### PR DESCRIPTION
**Pull request by duament**
#234 #387

#### Usage
See `ss-redir`: https://github.com/shadowsocks/shadowsocks-libev#transparent-proxy

### CODE REVIEW

File: proxy/redir/utils_freebsd.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package redir
+
+import (
+	"errors"
+	"net"
+)
+
+func dialUDP(network string, lAddr *net.UDPAddr, rAddr *net.UDPAddr) (*net.UDPConn, error) {
+	return nil, errors.New("FreeBSD not support yet")
```
Comment (Dreamacro) [inline proxy/redir/utils_freebsd.go]: https://golang.org/pkg/go/build/#hdr-Build_Constraints would reduce utils_{platfom} file

File: proxy/listener.go [hunk 2]
```diff
@@ -141,6 +150,11 @@ func ReCreateRedir(port int) error {
 		return err
 	}
 
+	redirUDPListener, err = redir.NewRedirUDPProxy(addr)
```
Comment (Dreamacro) [inline proxy/listener.go]: it seems that always return err on macOS.

File: proxy/redir/udp.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+func NewRedirUDPProxy(addr string) (*RedirUDPListener, error) {
+	l, err := net.ListenPacket("udp", addr)
+	if err != nil {
+		return nil, err
+	}
+
+	rl := &RedirUDPListener{l, addr, false}
+
+	c, ok := l.(*net.UDPConn)
+	if !ok {
```
Comment (Dreamacro) [inline proxy/redir/udp.go]: type check not necessary

File: proxy/redir/udp.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	l.closed = true
+	return l.PacketConn.Close()
+}
+
+func (l *RedirUDPListener) Address() string {
+	return l.address
+}
+
+func handleRedirUDP(pc net.PacketConn, buf []byte, addr *net.UDPAddr, origDst *net.UDPAddr) {
+	var origDstAddr net.Addr = origDst
```
Comment (Dreamacro) [inline proxy/redir/udp.go]: `net.Addr` is an interface, so golang could automate handle it

**Comment by duament** [comment]
> Is this pr have already tested on your device?

Yes, it works fine on my laptop running Arch Linux.

File: proxy/listener.go [hunk 5]
```diff
@@ -152,7 +153,7 @@ func ReCreateRedir(port int) error {
 
 	redirUDPListener, err = redir.NewRedirUDPProxy(addr)
 	if err != nil {
-		return err
+		log.Debugln("Failed to start Redir UDP Listener: %s", err)
```
Comment (Dreamacro) [inline proxy/listener.go]: maybe warning is better

**Comment by Dreamacro** [comment]
@duament run `go fmt ./...` at the root directory. And LGTM on code 👍

**Comment by Dreamacro** [comment]
Thanks for your contribution. As for TCP chaining, I have to speed the time to review.

**Pull request merged**